src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -779,6 +780,379 @@ BOOST_AUTO_TEST_CASE(GradientLSTMLayerTest)
+    }
+
+    ~GradientFunction()
+    {
+      delete model;
+    }
+
+    double Gradient(arma::mat& gradient) const
+    {
+      arma::mat output;
+      double error = model->Evaluate(model->Parameters(), 0);
+      model->Gradient(model->Parameters(), 0, gradient);
+      return error;
+    }
+
+    arma::mat& Parameters() { return model->Parameters(); }
+
+    RNN<NegativeLogLikelihood<> >* model;
+    arma::mat input, target;
+  } function;
+
+  BOOST_REQUIRE_LE(CheckGradient(function), 1e-4);
+}
+
+/**
+ * GRU layer manual forward test.
+ */
+BOOST_AUTO_TEST_CASE(ForwardGRULayerTest)
+{
+  GRU<> gru(3, 3, 5);
+
+  // Initialize the weights to all ones.
+  NetworkInitialization<ConstInitialization>
+    networkInit(ConstInitialization(1));
+  networkInit.Initialize(gru.Model(), gru.Parameters());
+
+  // Provide input of all ones.
+  arma::mat input = arma::ones(3, 1);
+  arma::mat output;
+
+  gru.Forward(std::move(input), std::move(output));
+
+  // Compute the z_t gate output.
+  arma::mat expectedOutput = arma::ones(3, 1);
+  expectedOutput *= -4;
+  expectedOutput = arma::exp(expectedOutput);
+  expectedOutput = arma::ones(3, 1) / (arma::ones(3, 1) + expectedOutput);
+  expectedOutput = (arma::ones(3, 1)  - expectedOutput) % expectedOutput;
+
+  // For the first input the output should be equal to the output of
+  // gate z_t as the previous output fed to the cell is all zeros.
+  BOOST_REQUIRE_LE(arma::as_scalar(arma::trans(output) * expectedOutput), 1e-2);
+
+  expectedOutput = output;
+
+  gru.Forward(std::move(input), std::move(output));
+
+  double s = arma::as_scalar(arma::sum(expectedOutput));
+
+  // Compute the value of z_t gate for the second input.
+  arma::mat z_t = arma::ones(3, 1);
+  z_t *= -(s + 4);
+  z_t = arma::exp(z_t);
+  z_t = arma::ones(3, 1) / (arma::ones(3, 1) + z_t);
+
+  // Compute the value of o_t gate for the second input.
+  arma::mat o_t = arma::ones(3, 1);
+  o_t *= -(arma::as_scalar(arma::sum(expectedOutput % z_t)) + 4);
+  o_t = arma::exp(o_t);
+  o_t = arma::ones(3, 1) / (arma::ones(3, 1) + o_t);
+
+  // Expected output for the second input.
+  expectedOutput = z_t % expectedOutput + (arma::ones(3, 1) - z_t) % o_t;
+
+  BOOST_REQUIRE_LE(arma::as_scalar(arma::trans(output) * expectedOutput), 1e-2);
+}
+
+/**
+ * MemoryHead manual forward test.
+ */
+BOOST_AUTO_TEST_CASE(ForwardMemoryHeadTest)
+{
+  arma::mat memory = arma::ones(5, 5);
+  arma::mat input = arma::ones(5, 1);
+
+  MemoryHead<> mh(5, 5, 5, 5);
+
+  NetworkInitialization<ConstInitialization>
+      networkInit(ConstInitialization(1));
+  networkInit.Initialize(mh.Model(), mh.Parameters());
+
+  arma::mat weights;
+  mh.ForwardWithMemory(std::move(input), std::move(memory), std::move(weights));
+
+  // approximate k_t gate output
+  arma::mat k_t = arma::ones<arma::mat>(5, 1);
+
+  double b_t = 6.0;
+  double g_t = 1.0;
+  double gamma_t = 6.0;
+
+  arma::mat cosSimilarity = arma::normalise(memory, 1) * k_t /
```

Comment:
Wondering if we could use https://github.com/mlpack/mlpack/blob/master/src/mlpack/core/kernels/cosine_distance.hpp here.